### PR TITLE
applyconfig: handle nonexisting namespaces in server-side dry-run

### PR DESCRIPTION
**Pull request by petr-muller**
This is an unpolished shot to make `applyconfig` handle nonexistent namespaces in dry-runs better. Currently, server-side dry-runs (default on non-api.ci clusters) fails on PRs that add manifests for both the namespace and the resources in it. Server-side dry-run will fail for the resources, because they are applied into a namespace that does not exist on the cluster yet, but the actual non-dry run would succeed because the namespace manifest would be applied previously.

The solution space has several kinks:
- We could ask people to add namespaces in separate PRs, to be merged first (works but terrible UX)
- We could ignore these errors, but then we lose all validation for newly added manifests (can cause triage alerts)
- We could proactively create the missing namespaces and retry. This gets us validation but would create even mismatching (really nonexistent namespaces) namespaces, and we would need to figure out how to clean up.

This PR implements the third option - if a manifest fails to apply because of a missing namespace, it creates the namespace and retries, until no more missing namespace messages occur or the apply passes. During the run, it bookkeeps which namespaces would be created if the run was not dry. If a manifest needed to create a missing namespace to be able to pass dry-run, `applyconfig` compares it with a set of namespaces that would be created if the run was not dry, and emits a failure if the assumed namespace would not be created.

This change makes applyconfig analyses the output of oc apply and
searches for messages that indicate that the apply failed because a
namespace was missing.  Then it tries to recover by creating these
missing namespaces and retrying the failed oc apply.  These namespaces
are created annotated for ci-ns-ttl-controller to reap it so unmerged
PRs do not clutter the cluster. If a manifest with the namespace is
actually applied (post-merge), the annotations are cleaned up.

/cc @stevekuznetsov @alvaroaleman 
/cc @openshift/openshift-team-developer-productivity-test-platform

### CODE REVIEW

**Comment by alvaroaleman** [comment]
@petr-muller re cleanup: Make the creation in dry run add a label, it will then automatically get removed by the non-dry creation (unless its in the manifest). Then make the ns-ttl conroller remove namespaces with that label oder than some grace period. WDYT?

**Comment by petr-muller** [comment]
That sounds good! I assumed I'd get the same behavior with labels as for annotations but haven't tried. In that case I think making ns-ttl controller cooperate is really the best option.

**Comment by alvaroaleman** [comment]
I should have fully read the your pr body, the behavior is the same for annotations and labels. But I can not confirm what you wrote above, they get removed when removed from the manifest as expected:
```
$ k apply -f ns.yaml -oyaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{"test":"value"},"creationTimestamp":null,"name":"alvaro-test"}}
    test: value
  creationTimestamp: "2021-01-22T18:01:52Z"
  name: alvaro-test
  resourceVersion: "451731185"
  selfLink: /api/v1/namespaces/alvaro-test
  uid: 3a320d9b-1634-48b0-9f0d-f70b360ed872
spec:
  finalizers:
  - kubernetes
status:
  phase: Active

[2021-01-22-13:02]:[app.ci@ci][1]/tmp/tmp.3M9YkoebeK
$ k apply -f ns.yaml -oyaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Namespace","metadata":{"annotations":{},"creationTimestamp":null,"name":"alvaro-test"}}
    openshift.io/sa.scc.mcs: s0:c38,c32
    openshift.io/sa.scc.supplemental-groups: 1001470000/10000
    openshift.io/sa.scc.uid-range: 1001470000/10000
  creationTimestamp: "2021-01-22T18:01:52Z"
  name: alvaro-test
  resourceVersion: "451731488"
  selfLink: /api/v1/namespaces/alvaro-test
  uid: 3a320d9b-1634-48b0-9f0d-f70b360ed872
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

**Comment by petr-muller** [comment]
@alvaroaleman that's not the behavior I'm seeing thought, I wonder what are we doing differently? How does your manifest look like?

```console
$ appci create ns muller-testing --as system:admin
namespace/muller-testing created

$ appci get ns muller-testing -o jsonpath='{.metadata.annotations.test}'

$ appci annotate namespace muller-testing "test=value" --as system:admin
namespace/muller-testing annotated

$ appci get ns muller-testing -o jsonpath='{.metadata.annotations.test}'
value⏎                                                                                                                                                                                                                                                                                        

$ cat ns.yaml 
apiVersion: v1
kind: Namespace
metadata:
  name: muller-testing

$ appci apply -f ns.yaml --as system:admin
Warning: resource namespaces/muller-testing is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by oc apply. oc apply should only be used on resources created declaratively by either oc create --save-config or oc apply. The missing annotation will be patched automatically.
namespace/muller-testing configured

$ appci get ns muller-testing -o jsonpath='{.metadata.annotations.test}'
value
```

**Comment by alvaroaleman** [comment]
@petr-muller the difference is that you use `kubectl annotate` which doesn't use the `apply` log. Just using `apply` twice will give you the behavior I demonstrated.

The reason is that the first `apply` will add the annotation to the `kubectl.kubernetes.io/last-applied-configuration` annotation, the second apply will then see it there and on the live object but not on the manifest and remove it.

`kubectl` considers all fields that have not been created by itself (tracked via the annotation) as `unmanaged` and will never remove them because they are not in the manifest (but would add them and start considering them managed once they are in the manifest).

**Comment by petr-muller** [comment]
I see. That's a bit unfortunate - I wanted to use `create` and `annotate` to create the namespace in `applyconfig` as it only has the missing NS name, not the full manifest (the NS can be one of the items in a list or  something). This means that for this to work, `applyconfig` would need to dump a namespace manifeste with an annotation and then apply it. I guess it's feasible.

File: pkg/nsttl/const.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package nsttl
```
Comment (alvaroaleman) [inline pkg/nsttl/const.go:1]: Can we put this below `pkg/api`, please?

**Review by alvaroaleman — COMMENTED** [review]
>  If a manifest with the namespace is
actually applied (post-merge), the annotations are cleaned up

I don't see any logic for that, am I blind?

**Comment by petr-muller** [comment]
```console
$ cat /home/afri/Temporary/applyconfig-namespace-test/1/*
apiVersion: v1
kind: Namespace
metadata:
  name: muller-testing-namespace
kind: List
apiVersion: v1
items:
- apiVersion: authorization.openshift.io/v1
  kind: RoleBinding
  metadata:
    name: admin
    namespace: muller-testing-namespace
  roleRef:
    name: tada
  subjects:
  - kind: UberUser
    name: petr-muller

$ go run ./cmd/applyconfig/ --config-dir ~/Temporary/applyconfig-namespace-test/1/
INFO[0000] Detecting dry run capabilities               
INFO[0000] oc version --output=json --as system:admin: OK 
INFO[0000] Selected dry run method                       method=server openshift-version=4.6.13
INFO[0000] Applying config in directory: /home/afri/Temporary/applyconfig-namespace-test/1/ 
INFO[0007] oc apply -f /home/afri/Temporary/applyconfig-namespace-test/1/00_namespaces.yaml -o name --as system:admin --dry-run=server --validate=true: OK 
INFO[0014] oc apply -f /home/afri/Temporary/applyconfig-namespace-test/1/others.yaml -o name --as system:admin --dry-run=server --validate=true: failed to apply
Error from server (NotFound): namespaces "muller-testing-namespace" not found 
INFO[0014] Running in server-side dry run mode, attempting to recover previous failure by providing missing namespaces 
INFO[0014] Temporarily creating missing namespace        missing-namespace=muller-testing-namespace
INFO[0020] oc apply -f - -o name --as system:admin: OK  
INFO[0027] oc apply -f /home/afri/Temporary/applyconfig-namespace-test/1/others.yaml -o name --as system:admin --dry-run=server --validate=true: failed to apply
Error from server: invalid kind for origin subject: "UberUser" 
INFO[0027] Running in server-side dry run mode, attempting to recover previous failure by providing missing namespaces 
ERRO[0027] Previous failure was not caused by missing namespaces, cannot recover 
ERRO[0027] Apply command failed (not recoverable)        output="Error from server: invalid kind for origin subject: \"UberUser\"\n"
ERRO[0027] There were failures while applying config     error="failed to apply config"
exit status 1

$ appci delete ns muller-testing-namespace --as system:admin
namespace "muller-testing-namespace" deleted

$ go run ./cmd/applyconfig/ --config-dir ~/Temporary/applyconfig-namespace-test/2/
INFO[0000] Detecting dry run capabilities               
INFO[0000] oc version --output=json --as system:admin: OK 
INFO[0000] Selected dry run method                       method=server openshift-version=4.6.13
INFO[0000] Applying config in directory: /home/afri/Temporary/applyconfig-namespace-test/2/ 
INFO[0007] oc apply -f /home/afri/Temporary/applyconfig-namespace-test/2/00_namespaces.yaml -o name --as system:admin --dry-run=server --validate=true: OK 
INFO[0009] oc apply -f /home/afri/Temporary/applyconfig-namespace-test/2/others.yaml -o name --as system:admin --dry-run=server --validate=true: failed to apply
Error from server (NotFound): namespaces "muller-testing-namespace" not found 
INFO[0009] Running in server-side dry run mode, attempting to recover previous failure by providing missing namespaces 
INFO[0009] Temporarily creating missing namespace        missing-namespace=muller-testing-namespace
INFO[0010] oc apply -f - -o name --as system:admin: OK  
INFO[0017] oc apply -f /home/afri/Temporary/applyconfig-namespace-test/2/others.yaml -o name --as system:admin --dry-run=server --validate=true: OK 
INFO[0017] Apply was successful only under namespace creation assumption  assumed-namespaces=muller-testing-namespace path=/home/afri/Temporary/applyconfig-namespace-test/2/others.yaml
INFO[0017] Success!                                     

$ appci get ns muller-testing-namespace -o json | jq .metadata.annotations
{
  "ci.openshift.io/delete-at": "2021-02-11T19:22:31Z",
  "ci.openshift.io/ttl.hard": "1h0m0s",
  "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"v1\",\"kind\":\"Namespace\",\"metadata\":{\"annotations\":{\"ci.openshift.io/ttl.hard\":\"1h0m0s\"},\"creationTimestamp\":null,\"name\":\"muller-testing-namespace\"},\"spec\":{},\"status\":{}}\n",
  "openshift.io/sa.scc.mcs": "s0:c48,c2",
  "openshift.io/sa.scc.supplemental-groups": "1002260000/10000",
  "openshift.io/sa.scc.uid-range": "1002260000/10000"
}
```

**Comment by petr-muller** [comment]
> I don't see any logic for that, am I blind?

@alvaroaleman this is implicit: namespace needs to persist only of the PR merges. The PR merges only if the dry run passed, and the dry-run passes only when there is a manifest for it somewhere. So when the PR merges, the NS manifest gets `oc apply`ed and the annotations are cleaned up.

**Comment by petr-muller** [comment]
I'll try to explain this better in the comment.

**Comment by droslean** [comment]
I think that since things are getting more complicated IMO we need to investigate other options as well. Can we consider making the tool to encode the YAML into objects and use the go clients to update the items if the [dryRun field](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#UpdateOptions) in meta options can support server-side dry run as well?

File: cmd/applyconfig/applyconfig.go [hunk 2]
```diff
@@ -173,14 +180,116 @@ func makeOcApply(kubeConfig, context, path, user string, dry dryRunMethod) *exec
...
-	out, err := c.runAndCheck(cmd, "apply")
-	logrus.WithField("output", string(out)).Info("Ran apply command")
-	return err
+var namespaceNotFound = regexp.MustCompile(`Error from server \(NotFound\): namespaces "(.*)" not found`)
+
+func inferMissingNamespaces(applyOutput []byte) sets.String {
+	var ret sets.String
+	for _, line := range strings.Split(string(applyOutput), "\n") {
+		line := strings.TrimSpace(line)
+		if matches := namespaceNotFound.FindStringSubmatch(line); len(matches) == 2 {
```
[the remark this answers is not in the document]
Comment (petr-muller) [reply cmd/applyconfig/applyconfig.go:189]: You have a perverse definition of "just" :D That decision is unrelated to this PR. I would not be against that, but I do not find value in investing in that effort.

I agree that parsing `oc` output kinda crossed the line where `applyconfig` was basically a smarter shellscript. I did not want to rewrite the tool entirely, it was using `oc` before so I just went with it. If someone wants to rewrite the internals to use pure go I'll happily review :)

File: cmd/applyconfig/applyconfig.go [hunk 3]
```diff
@@ -173,14 +180,116 @@ func makeOcApply(kubeConfig, context, path, user string, dry dryRunMethod) *exec
...
+	out, err := do()
+	for err != nil {
+		if retry := compensate(out); !retry {
+			logrus.WithField("output", string(out)).Errorf("Apply command failed (not recoverable)")
+			return namespaces, err
+		}
+		out, err = do()
+	}
+
+	for _, line := range strings.Split(string(out), "\n") {
```
[the remark this answers is not in the document]
Comment (petr-muller) [reply cmd/applyconfig/applyconfig.go:269]: Sure, I didn't think it's that complicated but more tests rarely hurt.

File: cmd/applyconfig/applyconfig.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (droslean) [reply cmd/applyconfig/applyconfig.go:189]: Yea it's fine for now. Sorry, my comment was too general and not related to this PR.

[the remark this answers is not in the document]
Comment (petr-muller) [reply cmd/applyconfig/applyconfig.go:189]: No need to apologize, your comment was totally valid! My comment came out too harsh.

I tried to explain what led me to doing this way (I would need to rewrite the whole thing and I thought it's not worth the effort).

**Review by droslean — APPROVED** [review]

**Pull request merged**